### ISSUE TITLE
Intellij IDEA module structure not quite right?

### ISSUE THREAD

**Issue by jordansamuels** [Netflix/zuul#29]
TL;DR: some of IDEA's cool static analysis features don't work right with zuul code.  The only solution I can come up with on my own may be too invasive, since it touches on separating test/prod code.

I am using Intellij IDEA 13.1.3 (latest) with the Zuul code, and my main goal is to use the IDE to browse/analyze the code (e.g.  `Find Usages`, `Navigate -> Implementations`, `Navigate -> Call Hierarchy`).  

There are two probably-related problems I'm having "out of the box" (clean git clone; `./gradlew idea`; open `zuul.ipr` in IDEA [with the Groovy and Gradle plugins active]):
1. Groovy files are not recognized - e.g. `Navigate -> Implementations` on `ZuulFilter` only finds the Java implementors, not any of the Groovy ones).
2. IDEA's ability to filter test code from production code is hampered by the fact that IDEA thinks all the code is test code.

These may be separate underlying issues; I'm including them both in this entry as I believe they are related enough to discuss together.

To solve #1 we can "repair" the module structures; I say repair because IDEA itself doesn't like the fact that src/main/java appears twice each module, once for test and once for non-test:

```
$ grep src/main/java */*.iml
zuul-core/zuul-core.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="false" />
zuul-core/zuul-core.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="true" />
zuul-netflix-webapp/zuul-netflix-webapp.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="false" />
zuul-netflix-webapp/zuul-netflix-webapp.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="true" />
zuul-netflix/zuul-netflix.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="false" />
zuul-netflix/zuul-netflix.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="true" />
zuul-simple-webapp/zuul-simple-webapp.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="false" />
zuul-simple-webapp/zuul-simple-webapp.iml:      <sourceFolder url="file://$MODULE_DIR$/src/main/java" isTestSource="true" />
```

Without resolving all of those (e.g. remove the "true" entries), any changes to the module structure will cause IDEA to complain (e.g. if you try to apply any change at all) as it doesn't expect duplicate entries.  Resolving all of them will satisfy it, and then Problem #1 above seems to go away.

Problem #2 is of course also related to the test/prod distinction.  My selfish knee-jerk solution would be to separate test code into src/test/\* and remove this one line from [build.gradle](https://github.com/Netflix/zuul/blob/master/build.gradle):

```
sourceSets.test.java.srcDir 'src/main/java'
```

However I am aware of [Ben Christensen's viewpoint on test/production code](http://benjchristensen.com/2011/10/23/junit-tests-as-inner-classes/), and that not all zuul developers use IDEA and/or value static code analysis as much as I do.  I also agree in spirit that code readability should trump tool concerns, if there is indeed an unavoidable tradeoff.

If this can be resolved somehow, I consider it a big win for IDEA users, since the Navigation/Inspection features are so powerful, and help developers come up the learning curve.  It's a good complement to learning via debug/breakpoints, which for `zuul-netflix-webapp` is a bit more non-trivial as it requires getting all the components up and running.

**Comment by manuelcorrea** [comment]
@mikeycohen  is it possible to run the tests with  './gradlew test'  ?  or how are you running the test suite in the filters ? 
https://github.com/Netflix/zuul/issues/32

**Comment by manuelcorrea** [comment]
got it working adding this lines in the build.gradle

test {
    testLogging {
        events "passed", "skipped", "failed"
    }
}

sourceSets {
    test {
        groovy {
            srcDir 'src/main/groovy/filters'
        }
    }
}
